### PR TITLE
Add Russian language

### PR DESCRIPTION
**Pull request by karpnv**
Add Russian yaml, class RUCharParser() and relative path support in __parse_item() 
Add "manifest_line_parser" argument and one more parser: __parse_open_stt()

### CODE REVIEW

File: nemo/collections/asr/models/ctc_models.py [hunk 1]
```diff
@@ -631,8 +631,6 @@ def _setup_transcribe_dataloader(self, config: Dict) -> 'torch.utils.data.DataLo
             'sample_rate': self.preprocessor._sample_rate,
             'labels': self.decoder.vocabulary,
             'batch_size': min(config['batch_size'], len(config['paths2audio_files'])),
-            'trim_silence': True,
-            'shuffle': False,
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:635]: Please revert the shuffle = False.

File: nemo/collections/asr/parts/manifest.py [hunk 2]
```diff
@@ -88,6 +113,8 @@ def __parse_item(line: str, manifest_file: str) -> Dict[str, Any]:
             f"Manifest file {manifest_file} has invalid json line structure: {line} without proper audio file key."
         )
     item['audio_file'] = expanduser(item['audio_file'])
+    if item['audio_file'][0] != "/":
```
Comment (titu1994) [inline nemo/collections/asr/parts/manifest.py:116]: Forward slash is not always the path seperator, use os.sep everywhere.

File: nemo/collections/asr/parts/manifest.py [hunk 3]
```diff
@@ -99,7 +126,9 @@ def __parse_item(line: str, manifest_file: str) -> Dict[str, Any]:
     if 'text' in item:
         pass
     elif 'text_filepath' in item:
-        with open(item.pop('text_filepath'), 'r') as f:
+        if item['text_filepath'][0] != "/":
+            item['text_filepath'] = join(manifest_path, item['text_filepath'])
```
Comment (titu1994) [inline nemo/collections/asr/parts/manifest.py:130]: Same comment as above

File: nemo/collections/asr/data/audio_to_text.py [hunk 4]
```diff
@@ -132,6 +132,7 @@ def __init__(
         bos_id: Optional[int] = None,
         eos_id: Optional[int] = None,
         pad_id: int = 0,
+        manifest_line_parser: Callable[[str, Optional[str]], Dict[str, Any]] = None,
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_text.py:135]: This should be Optional[Callable...]

File: nemo/collections/asr/data/audio_to_text_dataset.py [hunk 5]
```diff
@@ -45,6 +45,7 @@ def get_char_dataset(config: dict, augmentor: Optional['AudioAugmentor'] = None)
         normalize=config.get('normalize_transcripts', False),
         trim=config.get('trim_silence', False),
         parser=config.get('parser', 'en'),
+        manifest_line_parser=config.get('manifest_line_parser', None),
```
Comment (titu1994) [inline nemo/collections/asr/data/audio_to_text_dataset.py:48]: A Callable function can't really be part of a manifest - so how is this handled ?

File: nemo/collections/asr/parts/manifest.py [hunk 6]
```diff
@@ -63,6 +63,8 @@ def item_iter(
 
     if parse_func is None:
         parse_func = __parse_item
+    elif parse_func == "open_stt":
```
Comment (titu1994) [inline nemo/collections/asr/parts/manifest.py:66]: Add docstring about this new function support.

File: nemo/collections/asr/parts/collections.py [hunk 7]
```diff
@@ -190,7 +190,7 @@ def __init__(self, manifests_files: Union[str, List[str]], *args, **kwargs):
         """
 
         ids, audio_files, durations, texts, offsets, speakers, orig_srs = [], [], [], [], [], [], []
-        for item in manifest.item_iter(manifests_files):
+        for item in manifest.item_iter(manifests_files, manifest_line_parser):
```
Comment (titu1994) [inline nemo/collections/asr/parts/collections.py:193]: Add docstring for new optional callable

File: nemo/collections/asr/parts/manifest.py [hunk 8]
```diff
@@ -74,10 +76,33 @@ def item_iter(
...
+            item['text'] = f.read().replace('\n', '')
+
+    item['duration'] = float(line_list[2])
+
+    return item
 
 def __parse_item(line: str, manifest_file: str) -> Dict[str, Any]:
     item = json.loads(line)
-
+    manifest_path, _ = split(manifest_file)
```
Comment (titu1994) [inline nemo/collections/asr/parts/manifest.py:105]: Could you explain why this is necessary to change ?

File: nemo/collections/asr/parts/manifest.py [hunk 9]
```diff
@@ -74,10 +76,33 @@ def item_iter(
...
+    line_list = line.split(",")
+    assert len(line_list)==3, f"Incorrect fields number: {len(line_list)}"
+    item = dict(
+        offset=None,
+        speaker=None,
+        orig_sr=None,
+    )
+
+    item['audio_file'] = line_list[0]
+    if item['audio_file'][0] != "/":
```
Comment (titu1994) [inline nemo/collections/asr/parts/manifest.py:90]: Forward slash is not always the path seperator, use os.sep everywhere

File: nemo/collections/asr/parts/manifest.py [hunk 10]
```diff
@@ -74,10 +76,33 @@ def item_iter(
...
+        item['audio_file'] = join(manifest_path, item['audio_file'])
+
+    text_file = line_list[1]
+    if text_file[0] != "/":
```
Comment (titu1994) [inline nemo/collections/asr/parts/manifest.py:94]: Same as above

File: nemo/collections/asr/parts/manifest.py [hunk 11]
```diff
@@ -74,10 +76,33 @@ def item_iter(
 
                 yield item
 
+def __parse_open_stt(line: str, manifest_file: str) -> Dict[str, Any]:
+    manifest_path, _ = split(manifest_file)
+    line_list = line.split(",")
+    assert len(line_list)==3, f"Incorrect fields number: {len(line_list)}"
+    item = dict(
```
Comment (titu1994) [inline nemo/collections/asr/parts/manifest.py:83]: Please update the this item dict with values from the manifest (if if exists) - at least for offset

**Review by titu1994 — CHANGES_REQUESTED** [review]
Really cool ! Minor changes required.

**Comment by titu1994** [comment]
Could you sign the DCO commits via the instruction here - 
https://github.com/NVIDIA/NeMo/pull/2230/checks?check_run_id=2612023734

**Comment by titu1994** [comment]
Also seems test failed due to style formatting. Could you run `python setup.py style --scope nemo --fix`

File: nemo/collections/asr/parts/manifest.py (hunk 8 above)
Comment (karpnv) [reply nemo/collections/asr/parts/manifest.py:105]: Together with the line item['audio_file'] = join(manifest_path, item['audio_file']) you can hve inside manifest file relative paths to manifest position

File: nemo/collections/asr/parts/manifest.py (hunk 11 above)
Comment (karpnv) [reply nemo/collections/asr/parts/manifest.py:83]: OpenStt dataset does not have offset values. There are only 3 comma-separated fields: audio_file, text_file, duration. https://github.com/snakers4/open_stt  I move this function out to a separate file.

File: nemo/collections/asr/data/audio_to_text.py (hunk 4 above)
Comment (karpnv) [reply nemo/collections/asr/data/audio_to_text.py:136]: It should be the same as here:
https://github.com/NVIDIA/NeMo/blob/9c75c348bc8137717ca79a34a93c50028ffdfd38/nemo/collections/asr/parts/manifest.py#L35

File: nemo/collections/asr/models/ctc_models.py (hunk 1 above)
Comment (karpnv) [reply nemo/collections/asr/models/ctc_models.py:635]: Done

File: nemo/collections/asr/parts/manifest.py (hunk 2 above)
Comment (karpnv) [reply nemo/collections/asr/parts/manifest.py]: fixed by isabs()

File: nemo/collections/asr/parts/manifest.py (hunk 3 above)
Comment (karpnv) [reply nemo/collections/asr/parts/manifest.py:104]: fixed by isabs()

File: nemo/collections/asr/data/audio_to_text_dataset.py (hunk 5 above)
Comment (karpnv) [reply nemo/collections/asr/data/audio_to_text_dataset.py]: add hydra.utils.instantiate()

File: nemo/collections/asr/parts/collections.py (hunk 7 above)
Comment (karpnv) [reply nemo/collections/asr/parts/collections.py:194]: done

File: nemo/collections/asr/parts/manifest.py (hunk 9 above)
Comment (karpnv) [reply nemo/collections/asr/parts/manifest.py]: fixed by isabs()

File: nemo/collections/asr/parts/manifest.py (hunk 10 above)
Comment (karpnv) [reply nemo/collections/asr/parts/manifest.py]: fixed by isabs()

File: nemo/collections/asr/data/audio_to_text.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/data/audio_to_text.py:135]: Hmm the signature and how we use it is different. For now I guess its fine to leave as it is, it becomes way too long to add on Union[Optional[...], str].

File: nemo/collections/asr/data/audio_to_text_dataset.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/asr/data/audio_to_text_dataset.py]: Previously it was a string method embedded in NeMo yes? So now its a user modifiable hydra block?

File: nemo/collections/asr/parts/manifest.py (hunk 8 above)
Comment (titu1994) [reply nemo/collections/asr/parts/manifest.py:79]: Hmm generally we try for absolute paths, but makes sense.

File: nemo/collections/asr/parts/manifest.py (hunk 11 above)
Comment (titu1994) [reply nemo/collections/asr/parts/manifest.py]: Yep I know it doesnt have offset values, but the audio processing pipeline is capable of slicing off segments using the two keys of offset and duration even from whole files. So we can augment the manifest for speech classification in the future with some modifications to the manifest. Anyway it can be handled in the future I suppose.

**Comment by titu1994** [comment]
Could you rebase with the main branch? Seems incorrect branch was selected?

File: nemo/collections/asr/data/audio_to_text_dataset.py (hunk 5 above)
Comment (karpnv) [reply nemo/collections/asr/data/audio_to_text_dataset.py]: yes, now it is hydra block

**Comment by titu1994** [comment]
The asr parts was refactored in order to be modular, could you apply those changes in the conflicting files? Most of them just had path changes- but manifest and parsers now belong to nemo/collections/commons/parts instead of asr.

**Comment by titu1994** [comment]
Closed due to inactivity. Please reach out in order to resume work on it.
